src/onefetch/deps.rs

```diff
@@ -0,0 +1,122 @@
+        match *self {
+            PackageManager::Cargo => write!(f, "Cargo"),
+            PackageManager::GoModules => write!(f, "Go Modules"),
+            PackageManager::Npm => {
+                if std::path::Path::new("./yarn.lock").exists() {
+                    write!(f, "Yarn")
+                } else {
+                    write!(f, "Npm")
+                }
+            }
+            PackageManager::Pip => write!(f, "Pip"),
+        }
+    }
+}
+
+type DependencyParser = fn(&str) -> Option<i32>;
+
+pub struct Detector {
+    package_managers: HashMap<String, (DependencyParser, PackageManager)>,
+}
+
+// Package parsers go here. Parsers should take stirng contents and output a i32
+mod package_parsers {
+    use {regex::Regex, toml::Value};
+
+    pub fn cargo(contents: &str) -> Option<i32> {
+        let parsed = contents.parse::<Value>().unwrap();
+
+        Some(parsed["dependencies"].as_table().unwrap().len() as i32)
+    }
+
+    pub fn gomodules(contents: &str) -> Option<i32> {
+        let count = Regex::new(r"v[0-9]+").unwrap().find_iter(contents).count();
+
+        Some(count as i32)
+    }
+
+    pub fn npm(contents: &str) -> Option<i32> {
+        let parsed = json::parse(contents).unwrap();
+
+        Some(parsed["dependencies"].len() as i32)
+    }
+
+    pub fn pip(contents: &str) -> Option<i32> {
+        let count = Regex::new(r"(^[A-z]+)|(\n[A-z]+)").unwrap().find_iter(contents).count();
+
+        Some(count as i32)
+    }
+}
+
+impl Detector {
+    pub fn new() -> Self {
+        let mut package_managers: HashMap<String, (DependencyParser, PackageManager)> =
+            HashMap::new();
+
+        package_managers
+            .insert(String::from("Cargo.toml"), (package_parsers::cargo, PackageManager::Cargo));
+        package_managers.insert(
+            String::from("go.mod"),
+            (package_parsers::gomodules, PackageManager::GoModules),
+        );
+        package_managers
+            .insert(String::from("package.json"), (package_parsers::npm, PackageManager::Npm));
+        package_managers
+            .insert(String::from("requirements.txt"), (package_parsers::pip, PackageManager::Pip));
+
+        Self { package_managers }
+    }
+
+    pub fn get_deps_info(&self, dir: &str) -> Result<String> {
+        fn is_package_file(detector: &Detector, file_name: &str) -> bool {
+            detector.package_managers.iter().any(|(package_manager_file_name, _)| {
+                file_name.starts_with(package_manager_file_name)
+            })
+        }
+
+        let package_files = fs::read_dir(dir)
+            .chain_err(|| "Could not read directory")?
+            .filter_map(std::result::Result::ok)
+            .map(|entry| entry.path())
+            .filter(|entry| {
+                entry.is_file()
+                    && entry
+                        .file_name()
+                        .map(OsStr::to_string_lossy)
+                        .map(|s| is_package_file(&self, s.as_ref()))
```

Comment:
Since `is_package_file` ends up taking `&self` as an argument, why not `impl` it directly on `Detector` instead of defining it inside another `fn`?

Reply:
I'm not sure I understand. `is_package_file` doesn't take `self` as an argument? If I were to make it am associated function (static method?) I would have to call it with `onefetch::deps::DependencyDetector::is_package_file`, right?

Reply:
It's actually a class method :facepalm: so 6fdb61d